server/src/main/java/org/apache/iotdb/db/metadata/mtree/MTreeAboveSG.java

```diff
@@ -0,0 +1,533 @@
+      throws MetadataException {
+    CounterTraverser counter = new StorageGroupCounter(root, pathPattern);
+    counter.setPrefixMatch(isPrefixMatch);
+    counter.traverse();
+    return counter.getCount();
+  }
+
+  /**
+   * E.g., root.sg is storage group given [root, sg], return the MNode of root.sg given [root, sg,
+   * device], throw exception Get storage group node, if the give path is not a storage group, throw
+   * exception
+   */
+  public IStorageGroupMNode getStorageGroupNodeByStorageGroupPath(PartialPath path)
+      throws MetadataException {
+    IStorageGroupMNode node = getStorageGroupNodeByPath(path);
+    if (!node.getPartialPath().equals(path)) {
+      throw new MNodeTypeMismatchException(
+          path.getFullPath(), MetadataConstant.STORAGE_GROUP_MNODE_TYPE);
+    }
+
+    return node;
+  }
+
+  /**
+   * E.g., root.sg is storage group given [root, sg], return the MNode of root.sg given [root, sg,
+   * device], return the MNode of root.sg Get storage group node, the give path don't need to be
+   * storage group path.
+   */
+  public IStorageGroupMNode getStorageGroupNodeByPath(PartialPath path) throws MetadataException {
+    String[] nodes = path.getNodes();
+    if (nodes.length == 0 || !nodes[0].equals(root.getName())) {
+      throw new IllegalPathException(path.getFullPath());
+    }
+    IMNode cur = root;
+    for (int i = 1; i < nodes.length; i++) {
+      cur = cur.getChild(nodes[i]);
+      if (cur == null) {
+        break;
+      }
+      if (cur.isStorageGroup()) {
+        return cur.getAsStorageGroupMNode();
+      }
+    }
+    throw new StorageGroupNotSetException(path.getFullPath());
+  }
+
+  public List<IStorageGroupMNode> getInvolvedStorageGroupNodes(
+      PartialPath pathPattern, boolean isPrefixMatch) throws MetadataException {
+    List<IStorageGroupMNode> result = new ArrayList<>();
+    StorageGroupCollector<List<IStorageGroupMNode>> collector =
+        new StorageGroupCollector<List<IStorageGroupMNode>>(root, pathPattern) {
+          @Override
+          protected void collectStorageGroup(IStorageGroupMNode node) {
+            result.add(node);
+          }
+        };
+    collector.setCollectInternal(true);
+    collector.setPrefixMatch(isPrefixMatch);
+    collector.traverse();
+    return result;
+  }
+
+  /** Get all storage group MNodes */
+  public List<IStorageGroupMNode> getAllStorageGroupNodes() {
+    List<IStorageGroupMNode> ret = new ArrayList<>();
+    Deque<IMNode> nodeStack = new ArrayDeque<>();
+    nodeStack.add(root);
+    while (!nodeStack.isEmpty()) {
+      IMNode current = nodeStack.pop();
+      if (current.isStorageGroup()) {
+        ret.add(current.getAsStorageGroupMNode());
+      } else {
+        nodeStack.addAll(current.getChildren().values());
+      }
+    }
+    return ret;
+  }
+
+  /**
+   * Check whether the given path exists.
+   *
+   * @param path a full path or a prefix path
+   */
+  public boolean isPathExist(PartialPath path) {
```

Comment:
The name of this method is confusing.

Actually, this method cannot judge whether a path exists or not. It can only tell the invoker whether the SG of the path exists or not.

Reply:
Fixed. Already rename this method to ```isStorageGroupAlreadySet``` and comments have been updated.